main.go

```diff
@@ -111,123 +151,134 @@ func main() {
-			// timed out, this will catch us up to the current state of telemetry, allowing us to resume.
-			select {
-			case telemetryBytes := <-telemetryChan:
-				// We received telemetry
-				batch.AddTelemetry(lastRequestId, telemetryBytes)
-				util.Logf("We suspected a timeout for request %s but got telemetry anyway", lastRequestId)
-			default:
+			eventCounter++
+
+			if probablyTimeout {
+				// We suspect a timeout. Either way, we've gotten to the next event, so telemetry will
+				// have arrived for the last request if it's going to. Non-blocking poll for telemetry.
+				// If we have indeed timed out, there's a chance we got telemetry out anyway. If we haven't
+				// timed out, this will catch us up to the current state of telemetry, allowing us to resume.
+				select {
+				case telemetryBytes := <-telemetryChan:
+					// We received telemetry
+					batch.AddTelemetry(lastRequestId, telemetryBytes)
+					util.Logf("We suspected a timeout for request %s but got telemetry anyway", lastRequestId)
+				default:
+				}
 			}
-		}
 
-		if event.EventType == api.Shutdown {
-			if event.ShutdownReason == api.Timeout && lastRequestId != "" {
-				// Synthesize the timeout error message that the platform produces, and LLC parses
-				timestamp := eventStart.UTC()
-				timeoutSecs := eventStart.Sub(lastEventStart).Seconds()
-				timeoutMessage := fmt.Sprintf(
-					"%s %s Task timed out after %.2f seconds",
-					timestamp.Format(time.RFC3339),
-					lastRequestId,
-					timeoutSecs,
-				)
-				batch.AddTelemetry(lastRequestId, []byte(timeoutMessage))
-			} else if event.ShutdownReason == api.Failure && lastRequestId != "" {
-				// Synthesize a generic platform error. Probably an OOM, though it could be any runtime crash.
-				errorMessage := fmt.Sprintf("RequestId: %s A platform error caused a shutdown", lastRequestId)
-				batch.AddTelemetry(lastRequestId, []byte(errorMessage))
+			if event.EventType == api.Shutdown {
+				if event.ShutdownReason == api.Timeout && lastRequestId != "" {
+					// Synthesize the timeout error message that the platform produces, and LLC parses
+					timestamp := eventStart.UTC()
+					timeoutSecs := eventStart.Sub(lastEventStart).Seconds()
+					timeoutMessage := fmt.Sprintf(
+						"%s %s Task timed out after %.2f seconds",
+						timestamp.Format(time.RFC3339),
+						lastRequestId,
+						timeoutSecs,
+					)
+					batch.AddTelemetry(lastRequestId, []byte(timeoutMessage))
+				} else if event.ShutdownReason == api.Failure && lastRequestId != "" {
+					// Synthesize a generic platform error. Probably an OOM, though it could be any runtime crash.
+					errorMessage := fmt.Sprintf("RequestId: %s A platform error caused a shutdown", lastRequestId)
+					batch.AddTelemetry(lastRequestId, []byte(errorMessage))
+				}
+
+				pollLogServer(logServer, batch)
+				finalHarvest := batch.Close()
+				shipHarvest(ctx, finalHarvest, telemetryClient, invokedFunctionARN)
+				return eventCounter
```

Comment:
I think we need to shut down the log server before the final poll. That way, when the platform sends us logs too late, the platform will notice that we're not around to receive them. If we do the final poll first, then we drop the last logs.